cdap-common/src/main/java/co/cask/cdap/internal/io/ReflectionRowRecordReader.java

```diff
@@ -16,44 +16,123 @@
+        @Override
+        public String get(byte[] rowKey) {
+          return Bytes.toString(rowKey);
+        }
+      })
+      .build();
+  private final String rowFieldName;
+  private final RowKeyFunction rowKeyFunction;
 
-  public ReflectionRowRecordReader(Schema schema) {
-    super(schema, TypeToken.of(StructuredRecord.Builder.class));
-    Preconditions.checkArgument(schema.getType() == Schema.Type.RECORD, "Target schema must be a record.");
+  public ReflectionRowRecordReader(Schema schema, @Nullable String rowFieldName) {
+    super(schema, TypeToken.of(StructuredRecord.class));
+    this.rowFieldName = rowFieldName;
+    // if row field is given, make sure the type is a non-null simple type
+    if (rowFieldName != null) {
+      Schema.Field rowField = schema.getField(rowFieldName);
+      Schema.Type rowType = rowField.getSchema().getType();
+      Preconditions.checkArgument(rowField != null, "Row field not found in schema");
+      Preconditions.checkArgument(rowType != Schema.Type.NULL, "Row field cannot have null type.");
+      Preconditions.checkArgument(rowType.isSimpleType(),
+        "Row field must be a simple type (boolean, bytes, int, long, float, double, or string).");
+      this.rowKeyFunction = rowKeyFuctions.get(rowType);
+    } else {
+      this.rowKeyFunction = null;
+    }
   }
 
   @Override
-  public StructuredRecord.Builder read(Row row, Schema sourceSchema) throws IOException {
+  public StructuredRecord read(Row row, Schema sourceSchema) throws IOException {
     Preconditions.checkArgument(sourceSchema.getType() == Schema.Type.RECORD, "Source schema must be a record.");
     initializeRead(sourceSchema);
     StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    // if one of the fields should come from the row key, add it.
+    if (rowFieldName != null) {
+      builder.set(rowFieldName, rowKeyFunction.get(row.getRow()));
+    }
+
+    // go through the Row columns and add their values to the record
     try {
       for (Schema.Field sourceField : sourceSchema.getFields()) {
         String sourceFieldName = sourceField.getName();
         Schema.Field targetField = schema.getField(sourceFieldName);
-        if (targetField == null) {
+        // the Row may contain more fields than our target schema. Skip those fields that are not in the target schema,
+        // as well as the row key field since it comes from the row key and not the columns.
+        if (targetField == null || targetField.getName().equals(rowFieldName)) {
           advanceField();
           continue;
         }
         builder.set(sourceFieldName, read(row, sourceField.getSchema(), targetField.getSchema(), type));
       }
-      return builder;
+      return builder.build();
     } catch (Exception e) {
       throw propagate(e);
     }
   }
+
+  // translates a row key into some other object.
+  private interface RowKeyFunction<T> {
+    T get(byte[] rowKey);
```

Comment:
`get` is a bit confusing, can we call it `convert`, `translate`, or something similar?